### PR TITLE
vulkaninfo: Replace devsim with profile json output

### PR DESCRIPTION
**Pull request by christophe-lunarg**
Initial commit to make the output occur, will need additional refinement.

### CODE REVIEW

File: vulkaninfo/vulkaninfo.cpp
Comment (christophe-lunarg) [inline vulkaninfo/vulkaninfo.cpp:971]: The profile file is not generated for any version of the profile schema but for the profile schema API version that matches a Vulkan Header revision ideally.

This allows avoiding to require Vulkan capabilities that didn't exist for this Vulkan Header revision.

**Review by christophe-lunarg — COMMENTED** [review]
The capabilities section is only a container for capabilities. 

It's only the profile section that says what a profile is made of. Vulkan Profiles can be defined within a file so that a file may express multiple variants of a profile.

However the Vulkaninfo output a specific kind of Vulkan Profiles: device (and driver specific) profile so there will be a single profile per file, this said with the portability extensions this may be different.

Note as well that a profile also specific how a Vulkan capability should be used. So if the driver reports Vulkan 1.2.198, the generated profile file should not rely on Vulkan 1.3 core structures.

You can use https://www.jsonschemavalidator.net/ and schemas in https://schema.khronos.org/vulkan/ to check the files are generated correctly.

**Comment by christophe-lunarg** [comment]
Currently, we only have schema versions for each SDK revision we shipped since 1.1.106 but I think we should generate a schema for each Vulkan Header version we can.

It's probably not more than a day of work.

However, it raises the question of uploading Vulkan schemas for each Vulkan header revision. Do we follow the rythmn of Vulkan Header revisions, do we do it during SDK release?

File: vulkaninfo/vulkaninfo.cpp
Comment (charles-lunarg) [reply vulkaninfo/vulkaninfo.cpp:988]: So this should be changed such that the schema used has the same API version as the device?
Also is there going to be an official location for these schema files? IE currently i'm referencing this repo but that doesn't seem very official.

**Comment by lunarpapillo** [comment]
FYI, the last CI failure seems to be due to a CMake 3.19 bug.  I'm rerunning.

**Comment by charles-lunarg** [comment]
@christophe-lunarg I have gotten this PR into a good shape. It passes the schema and the vulkan-profiles repo can build with the output json file. The Portability Subset is handled like the `VP_LUNARG_desktop_portability_2021.json`, where it creates a separate capability group and profile which includes the baseline and portability stuff.

File: scripts/vulkaninfo_generator.py
Comment (aqnuep) [inline scripts/vulkaninfo_generator.py:96]: The generator script also disallow-lists the various DRM format modifier list structures for the same reason as `VkVideoProfilesKHR`, i.e. because they contain pointers to arrays which shouldn't be used in output structures. In order to stay consistent with the Vulkan-Profiles toolchain, you probably should do the same here so that you don't accidentally generate profile JSONs containing those disallow-listed structures, as they won't comply with the schema or work with the library generator.

**Review by aqnuep — APPROVED** [review]
The changes look fine in general, but it's impossible to tell if anything is missing or wrong just looking at the code, so in the end what matters is that this is covered with sufficient level of testing.

**Comment by charles-lunarg** [comment]
There are a couple of things which aren't solidified yet.

*  The contributors section. I left this blank for now, what should go there? LunarG? Khronos? Anyone whose ever edited vulkaninfo?
* I found on hardware that supports the video extension the VkFormats blocks would contain empty sections for `VkVideoProfileKHR`. This may be from what @aqnuep described, that some structs shouldn't be printed.
* A portability device will output the portability data (features, extension, properties) in a separate capabilities block and creates a separate profile for it. Is this an appropriate way to handle portability devices? (let me know if an example json should be uploaded).

File: vulkaninfo/vulkaninfo.cpp
Comment (ziga-lunarg) [inline vulkaninfo/vulkaninfo.cpp:44]: Could be const

File: vulkaninfo/vulkaninfo.cpp
Comment (ziga-lunarg) [inline vulkaninfo/vulkaninfo.cpp:48]: const

File: vulkaninfo/vulkaninfo.cpp
Comment (ziga-lunarg) [inline vulkaninfo/vulkaninfo.cpp:467]: Initializing with `{}` is better

File: vulkaninfo/vulkaninfo.cpp
Comment (ziga-lunarg) [inline vulkaninfo/vulkaninfo.cpp:642]: Not sure if I'm missing something, but I don't see where `sType` is set. `pNext` is set in `setup_format_properties2_chain`, but I would also like to see it set to `nullptr` here for clarity.

File: vulkaninfo/vulkaninfo.cpp
Comment (ziga-lunarg) [inline vulkaninfo/vulkaninfo.cpp:677]: const

File: vulkaninfo/vulkaninfo.cpp
Comment (ziga-lunarg) [inline vulkaninfo/vulkaninfo.cpp:745]: I always like to make references const

**Comment by aqnuep** [comment]
> I found on hardware that supports the video extension the VkFormats blocks would contain empty sections for VkVideoProfileKHR. This may be from what @aqnuep described, that some structs shouldn't be printed.

Yes, that may be related. We need to clarify intent on this one.

> A portability device will output the portability data (features, extension, properties) in a separate capabilities block and creates a separate profile for it. Is this an appropriate way to handle portability devices? (let me know if an example json should be uploaded).

I'd say that if we want to generate separate portability fallback profiles in the same profile JSON for each device then we might have to follow a similar pattern to the desktop portability profiles, i.e. have a separate capability set for the portability specific part and for stuff that cannot be in portability. This latter is important as e.g. some features like robust buffer access are not required in portability.

However, I'm not sure whether it's needed to generate a portability-compatible profile as well. That's open for debate I think.

File: vulkaninfo/vulkaninfo.cpp
Comment (christophe-lunarg) [reply vulkaninfo/vulkaninfo.cpp]: Yes same API version than the device.

The official location: https://schema.khronos.org/vulkan/

Comment (charles-lunarg) [reply vulkaninfo/vulkaninfo.cpp]: I can't use the same schema as the device version because it may not be on the website.
For example, MoltenVK uses 1.1.304.

File: vulkaninfo/vulkaninfo.cpp
Comment (charles-lunarg) [reply vulkaninfo/vulkaninfo.cpp:642]: `setup_format_properties2_chain` will set the pnext pointer. I changed the declaration to `VkFormatProperties2 format_props2{};` so it should be null'd out there.

File: scripts/vulkaninfo_generator.py
Comment (charles-lunarg) [reply scripts/vulkaninfo_generator.py:96]: The struct blacklist now contains the following.
```python
struct_blacklist = ['VkVideoProfilesKHR', 'VkVideoProfileKHR', 'VkDrmFormatModifierPropertiesListEXT', 'VkDrmFormatModifierPropertiesEXT', 'VkDrmFormatModifierPropertiesList2EXT']
```

**Comment by charles-lunarg** [comment]
> I'd say that if we want to generate separate portability fallback profiles in the same profile JSON for each device then we might have to follow a similar pattern to the desktop portability profiles, i.e. have a separate capability set for the portability specific part and for stuff that cannot be in portability.

This is what I have done currently. If running on a portability device, there will be a `macos-specific` capabilities object with the portability features, properties, and extension, while the `device` capabilities will omit those features/properties/extensions. Whether is necessary I will let others decide.

**Comment by charles-lunarg** [comment]
This PR is waiting on a final decision on how to handle the schema versioning. Currently the output uses a `profiles-1.x.y.json` but that has difficulties with systems that don't support the latest minor version with the latest patch version (such as MoltenVK).

**Comment by christophe-lunarg** [comment]
Looks great!

**Comment by charles-lunarg** [comment]
I'm going to merge this shortly (next day or two) so this is a reminder to any last minute reviews. I think the PR is in a good state. I don't recall there being any tasks left to do.

**Comment by christophe-lunarg** [comment]
Yes please!

**Comment by charles-lunarg** [comment]
Currently the output file is named `vulkaninfo.json`. Is this the default name we want? vulkan.gpuinfo.org prints the driver name and date in the filename.

**Comment by christophe-lunarg** [comment]
"vulkan.gpuinfo.org prints the driver name and date in the filename"

True, this is a good idea.

**Pull request merged**
